### PR TITLE
DPP-1478 make payment

### PR DESCRIPTION
**Pull request by carlbm**

### CODE REVIEW

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+using System.Linq;
+using NLog;
+using SFA.DAS.Payments.DCFS.Domain;
+using SFA.DAS.ProviderPayments.Calc.Refunds.Domain;
+using SFA.DAS.ProviderPayments.Calc.Refunds.Services.Dependiencies;
+using SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.ProviderPayments.Calc.Refunds.Services
+{
+    public class LearnerRefundProcessor : IProcessLearnerRefunds
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs:12]: ClassName and Interface name should they Match?

**Review by scottywakefield — APPROVED** [review]

File: src/Shared/SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure/Data/Entities/HistoricalPaymentEntity.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities
+{
+    public class HistoricalPaymentEntity : PaymentEntity
```
Comment (Pauljgraham) [inline src/Shared/SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure/Data/Entities/HistoricalPaymentEntity.cs:5]: Because Entities are very closely related to Data structures. It's generally considered a anti-pattern to for an entity to inherit from another entity

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            
+            foreach (var refundGroup in refundGroups)
+            {
+                foreach (var refund in refundGroup)
+                {
+                    var previousPaymentGroup = volatilePreviousPayments.ToLookup(x => new RefundGroup(x));
+                    var previousPaymentsForRefundGroup = new List<HistoricalPaymentEntity>();
+                    try
+                    {
+                        if (previousPaymentGroup.Contains(refundGroup.Key))
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs:46]: If no payments exist for a refund, then we shouldn't even try to process it, but we should log an error. This is definitely a different case from having negative payments.

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                        {
+                            previousPaymentsForRefundGroup = previousPaymentGroup[refundGroup.Key].ToList();
+                        }
+
+                        var newRefunds = ProcessRefund(refund, previousPaymentsForRefundGroup);
+                        refundPayments.AddRange(newRefunds);
+                        volatilePreviousPayments.AddRange(newRefunds.Select(x => new HistoricalPaymentEntity(x, refund)));
+                    }
+                    catch(ApplicationException)
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs:55]: We are using this exception for a very particular case, ApplicationException Is invalid and it makes sense to create a custom exception (if we decide to do this in this way)

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+            List<HistoricalPaymentEntity> previousPayments)
+        {
+            var refundPayments = new List<PaymentEntity>();
+
+            var amountToRefund = refund.AmountDue;
+            var amountRefunded = 0m;
+            var month = refund.DeliveryMonth;
+            var year = refund.DeliveryYear;
+
+            while (amountToRefund < amountRefunded)
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs:81]: It's quite complicated logic here. It's quite hard to follow. Is there a simple why to do this?

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+            {
+                var stillToRefund = amountToRefund - amountRefunded;
+                var paymentsForPeriod = previousPayments
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs:84]: This variable paymentforPeriod is only used in the ProcessRefundForPeriod() it should exist there as you also are passing in the filter values

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                    .Where(x => x.DeliveryMonth == month && x.DeliveryYear == year)
+                    .ToList();
+                var newRefunds = ProcessRefundForPeriod(stillToRefund, year, month, paymentsForPeriod, refund);
+                amountRefunded += newRefunds.Sum(x => x.Amount);
+                refundPayments.AddRange(newRefunds);
+            
+                if (month == 8)
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs:91]: This type of logic exist all over the system, to me it would make a lot of sense to turn this type of logic into a type and overload the + and - operators (but probably not for now). It would need to be in shared library

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+            return refundPayments;
+        }
+
+        private decimal TotalForFundingSource(List<HistoricalPaymentEntity> payments, FundingSource fundingSource)
+        {
+            var total = payments.Where(x => x.FundingSource == fundingSource).Sum(x => x.Amount);
+            if (total < 0)
+            {
+                throw new ApplicationException("Funding source total is negative");
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs:146]: ApplicationException should not be used and we need a custom exception

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/RefundGroup.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using SFA.DAS.Payments.DCFS.Domain;
+using SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.ProviderPayments.Calc.Refunds.Domain
+{
+    class RefundGroup : IEquatable<RefundGroup>
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Domain/RefundGroup.cs:7]: IS this not the Key Identifier for the group?

**Review by Pauljgraham — COMMENTED** [review]
This PR doesn't appear to have the unit tests 

AndThePaymentsAreSuffientToCoverTheRefund

just wondering how this were added

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests/ServiceTests/GivenALearnerRefundProcessor/AndThereArePreviousPaymentsForTheMonth.cs [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+using SFA.DAS.ProviderPayments.Calc.Refunds.Services;
+using SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests.Utilities;
+using SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests.Utilities.Extensions;
+using SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests.Utilities.TestHelpers;
+using SFA.DAS.ProviderPayments.Calc.Shared.Infrastructure.Data.Entities;
+
+namespace SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests.ServiceTests.GivenALearnerRefundProcessor
+{
+    [TestFixture]
+    public class AndThereArePreviousPaymentsForTheMonth
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests/ServiceTests/GivenALearnerRefundProcessor/AndThereArePreviousPaymentsForTheMonth.cs:15]: I really like this layout, can we as a sub team agree a standard going forward on how we write unit tests. We have several different styles floating around. One question if the Folder is GivenA.... then shouldn't each class be When.... and the sub classes be And....

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests/ServiceTests/GivenALearnerRefundProcessor/AndThereArePreviousPaymentsForTheMonth.cs [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+
+                actual.Should().HaveCount(9);
+            }
+        }
+
+        [TestFixture]
+        public class AndThePaymentsAreNotSufficientToCoverTheRefund
+        {
+            [Test]
+            [CreateMatchingRefundsAndPayments(hasMatchingPastPayments: false)]
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests/ServiceTests/GivenALearnerRefundProcessor/AndThereArePreviousPaymentsForTheMonth.cs:37]: I really like this, but is it doing the same thing as AutoFixture (but in a different way)?

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests/ServiceTests/GivenALearnerRefundProcessor/AndThereArePreviousPaymentsForTheMonth.cs [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+            public void AndThereAreNoPastPayments_ThenTheRefundsIsZero(
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests/ServiceTests/GivenALearnerRefundProcessor/AndThereArePreviousPaymentsForTheMonth.cs:38]: Shouldn't we be logging an error for this case?

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        {
+            _logger = logger;
+        }
+
+        public List<PaymentEntity> ProcessRefundsForLearner(
+            List<RequiredPaymentEntity> refunds,
+            List<HistoricalPaymentEntity> previousPayments)
+        {
+            var refundGroups = refunds.ToLookup(x => new RefundGroup(x));
+            var volatilePreviousPayments = new List<HistoricalPaymentEntity>(previousPayments);
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds/Services/LearnerRefundProcessor.cs:34]: Can we call this copyOfPreviousPayments as it's not volatile in the c# sense. You are not changing values only adding to a list

File: src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests/ServiceTests/GivenALearnerRefundProcessor/AndThereArePreviousPaymentsForTheMonth.cs [hunk 14]
```diff
@@ -0,0 +1,373 @@
...
+            {
+                var data = RefundGenerator.Generate(numberOfRefunds: 2, refundAmount: -900, paymentAmount: 500);
+
+                var refundOne = data.Refunds[0];
+                refundOne.ApprenticeshipContractType = 1;
+                var refundTwo = data.Refunds[1];
+                refundTwo.ApprenticeshipContractType = 2;
+                refundTwo.AmountDue = -1200;
+
+                data.AssociatedPayments[0].ApprenticeshipContractType = 1;
```
Comment (Pauljgraham) [inline src/PeriodEnd/Refunds/SFA.DAS.ProviderPayments.Calc.Refunds.UnitTests/ServiceTests/GivenALearnerRefundProcessor/AndThereArePreviousPaymentsForTheMonth.cs:180]: This is a real side issue (but we have enums for the FundingSource and TransactionType), why not for teh apprenticeshipContractType (we don't need to resolve this now)
